snapcraft/services/lifecycle.py

```diff
@@ -36,5 +65,75 @@ def setup(self) -> None:
             base=project.get_effective_base(),
             extra_build_snaps=project.get_extra_build_snaps(),
         )
-
         super().setup()
+
+    @overrides
+    def run(self, step_name: str | None, part_names: list[str] | None = None) -> None:
+        super().run(step_name, part_names)
+
+        enable_manifest = utils.strtobool(os.getenv("SNAPCRAFT_BUILD_INFO", "n"))
+        if enable_manifest:
+            self._manifest = self._generate_manifest()
```

Comment:
should this be done for any `step_name`? Or only `prime`?

Reply:
You are right, and the same applies to mandatory project variable verification. It can't run in a lifecycle epilogue callback because it runs at the end of any lifecycle run regardless of target step, and the post-prime callback runs for each part. It seems that specifically testing for the target step is necessary?

Reply:
Updated craft-application in https://github.com/canonical/craft-application/pull/257 to make manifest data generation (and mandatory adoptable field checking) to happen in the package service.